AFRL CSV to JSON (new API compatible)

bearmit opened the pull request:
**this PR is a draft until we have the authorization to publish the data, but code can be reviewed from now**

This PR contains the source code (typescript) to process AFRL dataset (CSV) and to produce a JSON compatible with the API.
Related to the task: https://trello.com/c/4eTnhJrp/29-transform-the-afrl-dataset

**Details:**

I explained how this works in a dedicated [README.md](https://github.com/C-Accel-CRIPT/criptscripts/tree/feat/afrl-to-new-api/scripts/typescript/afrl-to-new-api).

The goal of this first pass is to produce a JSON compatible (uploaded [here](https://drive.google.com/drive/folders/1YHXwVf3qscRCqt-pqqqUtXz_jQG22onC) ), but does not guaranty the data is translated 100% accurately.

Here are the problems I was facing, I found work around but I clearly not found "the solution":

- one_phase_direction cannot be stored in CRIPT (+one_phase_direction vocab is not allowed)
- the original script was searching materials from their "cas", I cannot do this, so it may create duplicates between the JSON and the existing database. **solution: we will use chemical_id search once implemented**
- The field Property.value does not accept a number, I had to convert it to string.
- To avoid local duplicated nodes, I used the technique Fatjon show me (using `uid`). It works but the JSON contains 2x more data. There is a possibility to reduce this by removing extra fields for the materials in the inventories, but I am not sure the backend will appreciate.

Feel free to add any comment / suggestions. Thanks.

bearmit committed:
afrl: first commit

bearmit committed:
feat: convert python function signatures to TS

bearmit committed:
fix: delete built file

bearmit committed:
docs: add comments and start the port

bearmit committed:
feat: wrap functions and global vars in class

bearmit committed:
feat: convert get_citation

bearmit committed:
fix typings *.d.ts => *.ts

bearmit committed:
feat: converted upload() (renamed load())

bearmit committed:
feat: convert get_polymer and smiles_to_BigSMILES

nh916 commented:
> First commit just to setup the typescript subproject to turn AFRL data into a JSON compatible with the new API. I just need a feed back regarding to the folder structure I proposed and where I want to add it. I won't merge until the feature is done, but I need your feedback about it.

I don't have any specific organization picked out for this repository when I was working on it, I just went with what I thought was intuitive and clean. I think the organization that you picked out could work, but I was thinking that if the script was only a single page then it might be easier to link to, such as the **bibliography page**
* [CRIPT Script page](https://criptscripts.org/scripts/bibliography%20creator.html)
* [Repository Page](https://github.com/C-Accel-CRIPT/criptscripts/tree/master/scripts/python_sdk_scripts/bibliography_creator)

If it needs to be longer, then that is no problem at all either. I would rather it be clean and longer than it be hard to read and shorter.

I think If we decide to put the current organization on CRIPT Script, then we would have to link to the GitHub repository directory where they can find it easily, like the **Simulation Procedure Script**
* [CRIPT Script page](https://criptscripts.org/scripts/simulation_procedure.html)
* [Repository Page](https://github.com/C-Accel-CRIPT/criptscripts/tree/master/scripts/python_sdk_scripts/simulation_procedure)

What are your thoughts?

bearmit committed:
convert get_mixture()

bearmit committed:
fix(get_citation): citation name always empty

bearmit committed:
refactor(get_citation): log message, remove commented code

bearmit committed:
feat: main loop finished

bearmit committed:
refactor(types): remove deprecated comments, sort fields

bearmit committed:
refactor(afrl): script name, remove deprecated

bearmit committed:
feat(afrl): write errors to a file, fix solvent hashmap

bearmit committed:
feat(afrl): generate a uid to avoid local duplicates

bearmit committed:
fix(afrl): smiles_to_bigsmiles

bearmit committed:
refactor(arfl): remove commented (and translated or useless) code

bearmit committed:
fix(afrl): get_citation's doi/author fields

bearmit committed:
fix(afrl): minor fixes

bearmit committed:
doc(afrl): add help to POST using curl

bearmit committed:
fix(afrl): missing "node" field for a solvent

bearmit committed:
fix(afrl): inventories

bearmit committed:
fix(afrl): collection missing "node" field

bearmit committed:
fix(afrl): hack to avoid backend issue on property.value

bearmit committed:
fix(afrl): remove "name" for Condition

The data model and API does not allow it.

bearmit committed:
fix(afrl): many fixes relative to "node" and "type" fields

bearmit committed:
doc(afrl): fix help with curl

bearmit committed:
fix(afrl): disable custom vocab

bearmit commented:
First version producing a JSON not rejected by the backend. Curl return success, but the backend does not seems to insert the data.

bearmit committed:
fix(afrl): "node" field type is now a static string

bearmit committed:
Revert "fix(afrl): "node" field type is now a static string"

This reverts commit b9f2268d817d94e928fa6dd406f32cc5477c1ff9.

bearmit committed:
feat(afrl): read from the CSV instead of the JSON

This will make easier the reimport form an other CSV.

bearmit committed:
refactor(afrl): arrow function factorized

bearmit committed:
fix(afrl): get_citation now logs existing_citation.reference.title

Was logging existing_citation.name, but this field is not allowed by the
API. However our typins allows it, not sure why.

bearmit committed:
fix(afrl): assign a uid not based on "cas"

bearmit committed:
fix(afrl): default values for numbers were wrong

bearmit committed:
fix(afrl): docs

bearmit commented on scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:
```diff
@@ -0,0 +1,631 @@
+
+/**
+ * This script is a typescript port of @see https://github.com/C-Accel-CRIPT/criptscripts/tree/master/scripts/python_sdk_scripts/AFRL
+ * The main difference is the data is already well structured to be used in TS (see ../data/data.ts).
```
This line is not correct. Data is now read from the CSV.

bearmit replied:
```suggestion
 * The main difference is the source data is now read from the original CSV and not a preprocessed *.js.
```

bearmit commented on scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:
```diff
@@ -0,0 +1,631 @@
+            model_version: MODEL_VERSION,
+            node: ['Project'],
+            collection: [this.collection],
+        } as IProject;
```
Should add a "material" field with an empty array.

bearmit commented on scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:
```diff
@@ -0,0 +1,631 @@
+        // Pull solvent from server
+        // Blocker: the script assumes the solvent already exists on the backend, but we probably don't have it.
+        /*
+        try {
```
Here, I skip the search because the API does not allow me to do it yet.

nh916 replied:
You may want to talk to @brili about this because I think he may want you to use an exact search from the API and replace all the material UUID with real UUID.

bearmit replied:
The commented code is searching material from a "cas" from CRIPT, I'll do once this is implemented yes.

bearmit commented on scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:
```diff
@@ -0,0 +1,631 @@
+        if (conc_vol_fraction && !isNaN(conc_vol_fraction)) {
+            mixture.property.push({
+                key: "conc_vol_fraction",
+                value: String(conc_vol_fraction), // FIXME: backend does not accept numbers,
```
number => string is incorrect, but backend requires it.

bearmit commented on scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:
```diff
@@ -0,0 +1,631 @@
+                // "components_relative" does not exist on new API, using "component" instead.         
+                component: [polymer],
+                citation,
+                node: ['Property'],
```
number => string is incorrect, but backend requires it.

bearmit commented on scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:
```diff
@@ -0,0 +1,631 @@
+            // There is no such field in the new API
+            // identifiers.append(cript.Identifier(key="preferred_name", value=name))
+            // Waiting for Brilant's answer I am using names instead.
+            mixture.names.push(name)
```
Not certain this value should be assigned to "names".

bearmit commented on scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:
```diff
@@ -0,0 +1,631 @@
+                component: [polymer],
+                citation,
+                node: ['Property'],
+                type: 'value',  // FIXME: is this correct from a chemist point of view?
```
Is this correct?

bearmit commented on scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:
```diff
@@ -0,0 +1,631 @@
+
+            const temp_cloud_property: IProperty = {
+                key: "temp_cloud",
+                value: String(temp_cloud), // FIXME: backend does not accept numbers
```
number => string is incorrect, but backend requires it.

bearmit commented on scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:
```diff
@@ -0,0 +1,631 @@
+                temp_cloud_property.condition.push({
+                    node: ['Condition'],
+                    key: "pressure",
+                    value: String(pressure), // FIXME: typings are wrong, we should be able to use a number
```
number => string is incorrect, but backend requires it.

bearmit commented on scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:
```diff
@@ -0,0 +1,631 @@
+
+                // FIXME: uncomment once backend accepts custom vocab (starts with a "+")
+                /*
+                temp_cloud_property.condition.push({
```
How do we deal with this condition key?

bearmit commented on scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:
```diff
@@ -0,0 +1,631 @@
+        this.polymers.forEach(assign_uid)
+
+        console.log('Loading data OK')
+        return this.project;
```
Major issue, no materials are added to the Project. They are only added to their respective Inventory.

bearmit commented on scripts/typescript/afrl-to-new-api/README.md:
```diff
@@ -0,0 +1,38 @@
+Step by step of what the script does on a macro level (for more details look at the source code).
+
+- Load the data (`./src/data/*.csv`) as an `Array<AFRLData>`.
+- For each data, we create 3 `Material`s (mixture, solvent, polymer) with some `Property`, `Condition` and `Citation` (`doi` is stored in a `Reference`) on it.
```
```suggestion
- For each CSV line (or object), we create 3 `Material`s (a mixture, a solvent, and a polymer) with some `Property`, `Condition` and `Citation` (`doi` is stored in a `Reference`) on it.
```

bearmit commented on scripts/typescript/afrl-to-new-api/README.md:
```diff
@@ -0,0 +1,38 @@
+
+This script was using Python originaly, last commit was from December 2022. I figured out the python version was existing **after** starting this task, that's why I used typescript.
+
+However, I couldn't us Python (no experience) and the PythonSDK (in development), so I decided to continue this way.
```
```suggestion
However, I couldn't use Python (no experience and culture) and the PythonSDK (still in development), so I decided to continue this way.
```

bearmit commented on scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:
```diff
@@ -0,0 +1,631 @@
+/**
+ * This script is a typescript port of @see https://github.com/C-Accel-CRIPT/criptscripts/tree/master/scripts/python_sdk_scripts/AFRL
+ * The main difference is the data is already well structured to be used in TS (see ../data/data.ts).
+ * The reason why data is a a AFRLData[] is because the legacy frontend (django app) had a JS file with the same data in it.
```
```suggestion
```

bearmit commented on scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:
```diff
@@ -0,0 +1,631 @@
+ * This script is a typescript port of @see https://github.com/C-Accel-CRIPT/criptscripts/tree/master/scripts/python_sdk_scripts/AFRL
+ * The main difference is the data is already well structured to be used in TS (see ../data/data.ts).
+ * The reason why data is a a AFRLData[] is because the legacy frontend (django app) had a JS file with the same data in it.
+ * Having a *.ts file is very usefull to have autocompletion and static typecheck. * 
```
```suggestion
```

bearmit committed:
Apply suggestions from code review

bearmit committed:
fix(afrl): missing "material" field in project.

Issue: cause the file to double size. We should find a way to reference
unsing an uid.

bearmit committed:
doc(afrl): remove deprecated comment

bearmit commented on scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:
```diff
@@ -0,0 +1,631 @@
+                /*
+                temp_cloud_property.condition.push({
+                    node: ['Condition'],
+                    key: "+one_phase_direction", // Not sure this will work, needs custom vocabulary (starts with a "+").
```
How do we deal with this condition key?

nh916 commented on scripts/typescript/afrl-to-new-api/README.md:
```diff
@@ -0,0 +1,38 @@
+
+# History
+
+This script was using Python originaly, last commit was from December 2022. I figured out the python version was existing **after** starting this task, that's why I used typescript.
```
I think it might be a good idea to link to the original Python AFRL script for easy find. There is also a typo in `"originaly"`

such as: 

> The script to upload this data to CRIPT had originally been a [python script](https://github.com/C-Accel-CRIPT/cript-ingestion-scripts/tree/master/afrl). The last commit was from December 2022.
> I figured out the python version existed **after** starting this task, which is why I wrote this using typescript.

bearmit replied:
> Love all the comments and clean code!

I actually kept a lot of the existing comments coming from the existing version ;)

nh916 commented:
@bearmit I am not sure if we can put this into [CRIPTScripts](https://criptscripts.org/) yet because I don't think we have the full legal clearance from Air Force Research Labs. I think @Ardi028 knows much more on this topic than I do

nh916 commented on scripts/typescript/afrl-to-new-api/src/index.ts:
```diff
@@ -0,0 +1,61 @@
+async function main(): Promise<number> {
+
+    console.log('=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-==-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=')
+    console.log('=-=-=-=-=-=-=-=-=-=-=-=-= AFRL CSV to JSON -=-=-=-=-=-==-=-=-=-=-=-=-=')
```
Love this!

nh916 commented on scripts/typescript/afrl-to-new-api/package.json:
```diff
@@ -0,0 +1,17 @@
+    "start": "npx ts-node src/index.ts",
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "author": "",
```
do you think it would be a good idea to add your name or CRIPT to the authors list?

bearmit replied:
I added my name + CRIPT + MIT. Also set the license to MIT too.

nh916 commented on scripts/typescript/afrl-to-new-api/package.json:
```diff
@@ -0,0 +1,17 @@
+{
+  "name": "afrl-transformation",
+  "version": "1.0.0",
+  "description": "",
```
do you think it would be a good idea to add a short description here?

nh916 commented:
Love all the comments and clean code!

bearmit commented:
I reopen because the source AFRL data (CSV file) is public.

bearmit committed:
docs(afrl): typo and link to Python script in History section

bearmit committed:
doc(afrl/package.json): add author, change license for MIT, description

bearmit commented on scripts/typescript/afrl-to-new-api/README.md:
```diff
@@ -0,0 +1,36 @@
+# AFRL transformation script
+
+The purpose of this node script is to convert AFRL dataset to a CRIPT compatible JSON.
+Related to: https://trello.com/c/4eTnhJrp/29-transform-the-afrl-dataset
```
This should not be merged :)
```suggestion
```

bearmit commented on scripts/typescript/afrl-to-new-api/package.json:
```diff
@@ -0,0 +1,17 @@
+  "main": "index.js",
+  "scripts": {
+    "start": "npx ts-node src/index.ts",
+    "test": "echo \"Error: no test specified\" && exit 1"
```
In order to produce qualitative data, we should use a testing framework to ensure some edge cases are well handled. I am thinking about:
- How an empty cell (string) is converted to number|string|other? (we want to control that)
- Can we find the location of the errors reported in *..errors.json inside the source CSV? (we store the csv row index, but I think we do not log it in errors messages)
- We should use a JSON validator (ajv) to validate the Project JSON before to send it to the backend
- We could automatize the POST within the typescript script.
- Is smiles_to_bigsmiles working correctly?

bearmit committed:
Update scripts/typescript/afrl-to-new-api/README.md

bearmit commented on scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:
```diff
@@ -0,0 +1,642 @@
+        // Try to get the existing
+        if (citation) {
+            // Note: previous implementation was using both mw_w, mw_d, and name as hash, but using unique_name seems more appropriate.
+            const existing_polymer = this.polymers.get(unique_name);
```
I am unsure of that now. Should I use the `cas` as keys in `this.polymer` map like I do for `this.solvents` ?
I case a change is required here, update line 281 accordingly.

nh916 replied:
From what I remember `CAS` was renamed to `chemical_id` for our platform

bearmit replied:
There is `cas` on `Material` schema. It is described in the `material_identifier_key ` controlled vocab too.

nh916 replied:
it should have been replaced with `chemical_id` because `CAS` is unavailable to our platform

bearmit replied:
You are correct. The vocabulary do not mention `cas` but do mention `chemical_id`. I'll still use `cas` in the code because the source data uses this concept.
Unfortunately I cannot store the `cas` as a `chemical_id` on a material with the current schema. Such a field does not exists. Since `cas` is present on the schema I fallback on this field temporarly:

![image](https://user-images.githubusercontent.com/111717478/230660195-fb616664-467d-45e7-9d31-56ff70e4638a.png)

bearmit commented on scripts/typescript/afrl-to-new-api/README.md:
```diff
@@ -0,0 +1,35 @@
+- Load the data (`./src/data/*.csv`) as an `Array<AFRLData>`.
+- For each CSV line (or object), we create 3 `Material`s (a mixture, a solvent, and a polymer) with some `Property`, `Condition` and `Citation` (`doi` is stored in a `Reference`) on it.
+- Each type of material is registered in a dedicated `Inventory`.
+- Inventories are store in a single `Collection` which is in a single `Project`.
```
```suggestion
- Inventories are stored in a unique `Collection` attached to a unique `Project`.
```

nh916 commented:
> > @bearmit I am not sure if we can put this into [CRIPTScripts](https://criptscripts.org/) yet because I don't think we have the full legal clearance from Air Force Research Labs. I think @Ardi028 knows much more on this topic than I do
> 
> Wait, is this repository public ??

yes it is

bearmit committed:
Update scripts/typescript/afrl-to-new-api/README.md

bearmit commented:
I am right now working on an other transformation (RCBC), I am based on this branch and had to refactor. I introduced a CriptJSON.stringify class (acting like JSON.stringify), to optimize the JSON file size (file size can be divided up to 4x).
See PR #6

bearmit commented:
I closed this PR, I reorganized my branches to have a clean history and a common branch for another script (in typescript too).
new PR: AFRL #8 
other script: RCBC #9
